add: Модуль лазерного целеуказателя на оружие

vovakons opened the pull request:
## Что этот ПР делает

Добавляет два новых модуля "Лазерный целеуказзатель (точка)" и "Лазерный целеуказатель (луч)".
Крепится на слот цевья (*большинство пистолетов, ппшки, дробовики*).
Работает только если держать оружие в руках, включается UI кнопкой или хоткеем Ctrl+R (*тот же хоткей что и у прицелов, если на оружии установлен прицел - у хоткея приоритет на прицел а не на целеуказатель*). Если убрать оружие с рук автоматически выключается.
Лазерный луч или точка примерно следует за указателем (*точность до тайла, увы ограничение движка, пока нет возможности сделать точность прям до указателя мыши*).
При активированном лазере дает 10 к точности (повышает шансы на попадание +10%) и уменьшает конус разброса на 25%.

## Почему это хорошо для игры

1. Новые модули на оружие, больше кастомизации, больше доходов НТ.
2. Можно аутентично угрожать применением табельного оружия с помощью лазерного луча или точки.
3. Совсем не нагружает, оптимизированно и переиспользует существующие данные для показа.

## Демонстрация изменений

https://github.com/user-attachments/assets/0cc64900-75c3-4bd7-88cb-71576ca3f424


## Тестирование
Протестировано на локалке. Присутствует небольшой лаг при движении по диагонали у луча, но не критично.

vovakons committed:
laser sight module

Добавлен новый модуль в слот подствольного крепежа - лазерный
целеуказатель.

vovakons committed:
laser sight component

Компонента отрисовки лазерного луча для нового модуля.

vovakons committed:
Merge branch 'master220' into laser-mod

vovakons committed:
split laser sight to point and ray

Разделение красной точки и лазерного луча.
Можно выбрать одну из модулей.

vovakons committed:
laser sight accuracy bonus only for enable mode

vovakons committed:
laser sight module in uplink

vovakons committed:
laser sight hotkey and vendings

vovakons committed:
scope priority for ctrl e hotkey

vovakons committed:
point sight random move in turf

vovakons committed:
random move for ray laser sight

vovakons commented:
Готово к ревью, локализацию тоже бы глянуть

littleboobs commented on code/modules/projectiles/gun_module.dm:
```diff
@@ -536,3 +536,76 @@
+
+/obj/item/gun_module/under/laser
+	name = "laser sight"
+	desc = "Модуль лазерного целеуказателя, позволяет увидеть куда вы целитесь с помощью лазерного луча. Повышает точность стрельбы и понижает разброс." //TODO описание хуйни - переделать
```
сильно

vovakons replied:
забыл про тудушку 😄

littleboobs commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+		point.pixel_x = clamp(point.pixel_x + rand(-1, 1), -move_range, move_range)
+		point.pixel_y = clamp(point.pixel_y + rand(-1, 1), -move_range, move_range)
+
+/datum/component/laser_sight/ray/proc/update_beam(atom/start, atom/end)
+	if(QDELETED(start) || QDELETED(end))
+		return
+	var/turf/start_turf = get_turf(start)
+	var/turf/end_turf = get_turf(end)
+	if(!start_turf || !end_turf)
+		return
+	//calculate transform
+	var/dx = (start_turf.x - end_turf.x) * ICON_SIZE_ALL + (start.pixel_x - end.pixel_x)
+	var/dy = (start_turf.y - end_turf.y) * ICON_SIZE_ALL + (start.pixel_y - end.pixel_y)
+	var/distance = sqrt(dx*dx + dy*dy)
+	var/angle = get_stable_angle(dx, dy)
+	var/matrix/trans = matrix()
+	trans.Translate(0, -ICON_SIZE_ALL/2)
+	//scale to distance
+	var/scale = max(distance / ICON_SIZE_ALL, 0)
+	trans.Scale(1, scale)
+	trans.Turn(angle)
+	animate(current_beam, transform = trans, time = 2)
+
+/datum/component/laser_sight/ray/proc/get_stable_angle(dx, dy)
+	var/angle = arctan(dy, dx)
+	return normalize_angle(angle)
+
+/datum/component/laser_sight/ray/proc/normalize_angle(angle)
+	while(angle > 90)
+		angle -= 360
+	while(angle < -90)
+		angle += 360
+	return angle
```
Мне кажется или это копипаст??

vovakons replied:
no, сам писал, чесна

littleboobs replied:
Тоже не ресолвай)

vovakons replied:
ну так в чем тут проблема? вроде все ок, даже комменты дал

littleboobs replied:
мне всё ещё не нравится копипаст, но Эшель тебе судья 🙏

littleboobs commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+
+/datum/action/toggle_laser_sight/Remove(mob/living/living)
+	sight.process_aim(living, FALSE)
+	..()
```
?
```suggestion
/datum/action/toggle_laser_sight/Remove(mob/living/living)
	sight.process_aim(living, FALSE)
	return ..()
```

littleboobs commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+	process_aim(user, FALSE)
+	action.Remove(user)
+	sight_target = null
+	return FALSE
```
```suggestion
/datum/component/laser_sight/proc/on_drop(datum/source, mob/user)
	SIGNAL_HANDLER

	process_aim(user, FALSE)
	action.Remove(user)
	sight_target = null
	return FALSE
```

littleboobs commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+		process_aim(user, FALSE)
+		action.Remove(user)
+		sight_target = null
+		return FALSE
```
```suggestion
/datum/component/laser_sight/proc/on_equip(datum/source, mob/user, slot)
	SIGNAL_HANDLER

	if(!(slot & ITEM_SLOT_HANDS))
		// If its not in their hands, disable laser, and remove the action button.
		process_aim(user, FALSE)
		action.Remove(user)
		sight_target = null
		return FALSE
```

littleboobs commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+
+
+
+// MARK: Ray
```
```suggestion
		point.forceMove(sight_target)

// MARK: Ray
```

littleboobs commented on code/modules/projectiles/gun_module.dm:
```diff
@@ -536,3 +536,76 @@
+/obj/item/gun_module/under/laser/on_detach(obj/item/gun/target_gun, mob/user)
+	var/datum/component/comp = target_gun.GetComponent(component_type)
+	QDEL_NULL(comp)
+	UnregisterSignal(target_gun, COMSIG_GUN_LASER_SIGHT_TOGGLE)
```
я, конечно, в коде шарю совсем, но это вызывает вопросы

vovakons replied:
нормально все

littleboobs replied:
Пускай Эшель ресолвнит, если всё нормально

littleboobs replied:
вот теперь норм

littleboobs reviewed (commented):
я не шарю, там где-то балуны спрятаны с вкл/выкл/перекл режим с точка на почка, если да, то ок, если нет, то)

vovakons committed:
laser sight module description

vovakons committed:
refiew fixes

littleboobs commented on code/modules/projectiles/gun_module.dm:
```diff
@@ -540,7 +540,7 @@
 /obj/item/gun_module/under/laser
 	name = "laser sight"
-	desc = "Модуль лазерного целеуказателя, позволяет увидеть куда вы целитесь с помощью лазерного луча. Повышает точность стрельбы и понижает разброс." //TODO описание хуйни - переделать
+	desc = "Модуль лазерного целеуказателя, крепится на крепление под стволом оружия. Повышает точность стрельбы и понижает разброс при стрельбе."
```
крепится на крепление
там же было где-то че-то с цевьём готовое в твоих же апдейтах

vovakons committed:
laser sight module description fix

vovakons committed:
Merge branch 'master220' into laser-mod

vovakons committed:
laser sight sprites and simple hand module

Обновлены спрайты для лазерного целеуказателя.
Добавлен новый модуль - прямая рукоятка, слабая версия угловой рукоятки.

dageavtobusnick committed:
Merge branch 'master220' into laser-mod

dageavtobusnick commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+
+/datum/component/laser_sight/point/on_disable_sight(mob/user)
+	QDEL_NULL(point)
+	if(rand_move_timer)
```
ранний ретурн

dageavtobusnick commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+		rand_move_timer = null
+
+/datum/component/laser_sight/point/on_update_sight(mob/user)
+	if(point.loc != sight_target)
```
ранний ретурн

dageavtobusnick commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+/datum/component/laser_sight/ray/proc/update_point(mob/user)
+	if(point.loc != sight_target)
+		point.forceMove(sight_target)
+	if(prob(50))
```
ранний ретурн

dageavtobusnick commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+		angle -= 360
+	while(angle < -90)
+		angle += 360
+	return angle
```
Сделай нормалайз угла как глобальный прок и запихни в математические хелперы. Это явно не должно быть персональным проком компонента

dageavtobusnick commented on code/datums/keybindings/mob.dm:
```diff
@@ -424,11 +424,14 @@
+	for(var/datum/action/action in human.actions)
+		if(istype(action, /datum/action/toggle_laser_sight))
+			action.Trigger()
+			break
```
Делай через сигнал. Плюсом 2 действия на 1 бинд звучит как херня

dageavtobusnick commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+		on_enable_sight(user)
+		update_sight_laser(user)
+		sight_timer = addtimer(CALLBACK(src, PROC_REF(update_sight_laser), user), 0.1, TIMER_STOPPABLE | TIMER_LOOP)
+	else
```
Ранний ретурн. Плюсом сам прок странноватый. Уверен, что это не 2 прока в 1?

dageavtobusnick commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,258 @@
+
+/datum/action/toggle_laser_sight/IsAvailable()
+	. = ..()
+	if(!. && sight)
```
ранний ретурн

dageavtobusnick commented on code/modules/projectiles/gun_module.dm:
```diff
@@ -521,18 +558,103 @@
+/obj/item/gun_module/under/laser/on_attach(obj/item/gun/target_gun, mob/user)
+	RegisterSignal(target_gun, COMSIG_GUN_LASER_SIGHT_TOGGLE, PROC_REF(laser_sight_toggle))
+	var/datum/component/laser_sight/comp = gun.AddComponent(component_type)
+	if(user.is_in_hands(target_gun))
```
ранний ретурн

dageavtobusnick commented on code/modules/projectiles/gun_module.dm:
```diff
@@ -521,18 +558,103 @@
+	RegisterSignal(target_gun, COMSIG_GUN_LASER_SIGHT_TOGGLE, PROC_REF(laser_sight_toggle))
+	var/datum/component/laser_sight/comp = gun.AddComponent(component_type)
+	if(user.is_in_hands(target_gun))
+		comp.on_equip(src, user, ITEM_SLOT_HANDS)
```
Это что такое? Сигнал делаем

vovakons committed:
Merge branch 'master220' into laser-mod

vovakons committed:
review fixes

vovakons committed:
gun attachment overlay bug fix

vovakons committed:
gun modules localization fix

dageavtobusnick commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,292 @@
+	angle = normalize_angle(angle)
+	get_angle()
+	var/matrix/trans = matrix()
+	trans.Translate(0, -ICON_SIZE_ALL/2)
```
```suggestion
	trans.Translate(0, -ICON_SIZE_ALL / 2)
```

dageavtobusnick commented on code/datums/components/laser_sight.dm:
```diff
@@ -0,0 +1,292 @@
+	//calculate transform
+	var/dx = (start_turf.x - end_turf.x) * ICON_SIZE_ALL + (start.pixel_x - end.pixel_x)
+	var/dy = (start_turf.y - end_turf.y) * ICON_SIZE_ALL + (start.pixel_y - end.pixel_y)
+	var/distance = sqrt(dx*dx + dy*dy)
```
```suggestion
	var/distance = sqrt(dx * dx + dy * dy)
```

littleboobs committed:
Merge branch 'master220' into laser-mod

vovakons committed:
review fixes

vovakons committed:
Merge branch 'laser-mod' of github.com:vovakons/Paradise into laser-mod
